packages/abtest/src/utils.js

```diff
@@ -0,0 +1,117 @@
+/**
+ * External dependencies
+ */
+import { recordEvent } from '@woocommerce/tracks';
+
+/**
+ * Internal dependencies
+ */
+import { CONTROL, EXPERIMENT, OPTION_NAME } from './constants';
+
+/**
+ * Get named key from localStorage.
+ *
+ * @param {string} name Key name.
+ *
+ * @return {string} - Value of named key.
+ */
+export const getCachedGroup = ( name ) => window.localStorage.getItem( name );
+
+/**
+ * Set localStorage named key to value.
+ *
+ * @param {string} name
+ * @param {string} value
+ */
+export const setCachedGroup = ( name, value ) =>
+	window.localStorage.setItem( name, value );
+
+/**
+ * Fetch option from backend and set localStorage accordingly.
+ *
+ * If option doesn't exist, then pick a group, update in the backend,
+ * set localStorage, and trigger a tracks event.
+ *
+ * If apiFetch fails, default to control.
+ *
+ * @param {string} name Experiment name.
+ * @param {number} size Perentage size of experimental group.
+ * @param {boolean} isOptionResolving Resolving state of selector.
+ * @param {Function} getOption Options data store selector.
+ * @param {Function} updateOptions Options data store dispatch.
+ *
+ * @return {string|null} - Group name (CONTROL or EXPERIMENT).
+ */
+export const getAndSetGroup = (
+	name,
+	size,
+	isOptionResolving,
+	getOption,
+	updateOptions
+) => {
+	try {
+		const option = getOption( OPTION_NAME );
+
+		if ( isOptionResolving ) {
+			return null;
```

Comment:
With the preloaded option removed, I can see this functioning as it should. Nothing is rendered until the option is returned, and that delay is only on the initial load. When the value is found in local storage, rendering is immediate 🎉 

I do think one improvement could be the addition of an optional prop on `ABTest` for a loading component. If, for whatever reason, I choose not to pre load the option, it would be nice to be able to supply an alternate component while this loads. The README would also be a good place to add some language describing how this works. What do you think?

Reply:
> I do think one improvement could be the addition of an optional prop on ABTest for a loading component. If, for whatever reason, I choose not to pre load the option, it would be nice to be able to supply an alternate component while this loads.

This was actually one of the primary reasons for adding the `onComplete` prop. It allows the parent to deal with any loading state as this may not always mean rendering a temporary component. Using a callback also removes some complexity from the package.

I can also see the benefit of explicitly passing a `loading` prop and moving the logic for this into the ABTest component though, and so I'm fine with either approach.

Let me know if you prefer one over the other and I'll go ahead and make the change, updating the README as well.

Reply:
I've gone ahead and added a section to the README showing how to deal with a loading state using the `onComplete` callback. If you prefer an explicit `loading` prop, let me know!